lib/mempool/addrindexer.js

```diff
@@ -0,0 +1,235 @@
+/*!
+ * mempool.js - mempool for bcoin
+ * Copyright (c) 2018-2019, the bcoin developers (MIT License).
+ * Copyright (c) 2014-2017, Christopher Jeffrey (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const {BufferMap} = require('buffer-map');
+const TXMeta = require('../primitives/txmeta');
+
+/**
+ * Address Indexer
+ * @ignore
+ */
+
+class AddrIndexer {
+  /**
+   * Create TX address index.
+   * @constructor
+   */
+
+  constructor() {
+    // Map of addr->entries.
+    this.index = new BufferMap();
+
+    // Map of txid->addrs.
+    this.map = new BufferMap();
+  }
+
+  reset() {
+    this.index.clear();
+    this.map.clear();
+  }
+
+  getKey(addr) {
+    const prefix = addr.getPrefix();
+
+    if (prefix < 0)
+      return null;
+
+    const raw = Buffer.allocUnsafe(1);
+    raw.writeUInt8(prefix);
+
+    return Buffer.concat([raw, addr.getHash()]);
+  }
+
+  /**
+   * Get transactions by address.
+   * @param {Address} addr
+   * @param {Object} options
+   * @param {Number} options.limit
+   * @param {Number} options.reverse
+   * @param {Buffer} options.after
+   */
+
+  get(addr, options = {}) {
+    const values = this.getEntries(addr, options);
+
+    const out = [];
+
+    for (const entry of values)
+      out.push(entry.tx);
+
+    return out;
+  }
+
+  /**
+   * Get transaction meta by address.
+   * @param {Address} addr
+   * @param {Object} options
+   * @param {Number} options.limit
+   * @param {Number} options.reverse
+   * @param {Buffer} options.after
+   */
+
+  getMeta(addr, options = {}) {
+    const values = this.getEntries(addr, options);
+
+    const out = [];
+
+    for (const entry of values) {
+      const meta = TXMeta.fromTX(entry.tx);
+      meta.mtime = entry.time;
+      out.push(meta);
+    }
+
+    return out;
+  }
+
+  /**
+   * Get entries by address.
+   * @param {Address} addr
+   * @param {Object} options
+   * @param {Number} options.limit
+   * @param {Number} options.reverse
+   * @param {Buffer} options.after
+   */
+
+  getEntries(addr, options = {}) {
+    const {limit, reverse, after} = options;
+    const key = this.getKey(addr);
+
+    if (!key)
+      return [];
+
+    const items = this.index.get(key);
+
+    if (!items)
+      return [];
+
+    let values = [];
+
+    // Check to see if results should be skipped because
+    // the after hash is expected to be within a following
+    // confirmed query.
+    const skip = (after && !items.has(after) && reverse);
+
+    if (skip)
+      return values;
+
+    if (after && items.has(after)) {
```

Comment:
To have similar interface, does not items need to be sorted array ?
for after/reverse to have similar effects as other indexers ?

Reply:
okay, I see. They are basically added in order so order will be same as they appeared -- similar to indexer.

Reply:
Yeah, exactly. If it's restored from persistence, I'm not sure of that order would be kept, however.

Reply:
Unfortunately they get sorted when stored, so that information will be lost.